allow_subclass won't let me pass a pointer to the superclass constructor in emscripten 4.0.22

Consider the trivial example code below, which compiles without error and functions as expected under `emcc 4.0.16`.  Under `emcc 4.0.22`, however, compilation fails with the error `static assertion failed due to requirement '!std::is_pointer<JJObserved *>::value': Implicitly binding raw pointers is illegal. Specify allow_raw_pointer<arg<?>>`, but I can't figure out how to correctly apply `allow_raw_pointer<arg<?>>` to the `constructor<JJObserved *>()` call.

I'm not sure if this is a bug/oversight in newer versions of Emscripten, or if I'm just ignorant of the necessary syntax to make this work.   If it's the latter, I would very much appreciate a hint regarding what I need to add to make this code compile again!

Thanks,
Jeremy

```#include <emscripten/bind.h>
#include <emscripten/wire.h>

using namespace emscripten;

class JJObserved
{
public:
   JJObserved() {/* empty */}
};

EMSCRIPTEN_BINDINGS(JJObserved)
{
   class_<JJObserved>("JJObserved");
}

class JJObserver
{
public:
   JJObserver(JJObserved * m) : _m(m) {/* empty */}

private:
   JJObserved * _m;
};

struct JJObserverWrapper : public wrapper<JJObserver>
{
   EMSCRIPTEN_WRAPPER(JJObserverWrapper);
};

EMSCRIPTEN_BINDINGS(JJObserverWrapper)
{
    class_<JJObserver>("JJObserver")
        .allow_subclass<JJObserverWrapper>("JJObserverWrapper", constructor<JJObserved *>());
}
```

For reference, here is the compile-time output I get from `emcc 4.0.22`:

```$ \make
emcc -I../../../libs/libmslicommon/include -I../../../submodules/muscle -I../../../submodules/Clipper2/emscripten_temp_install/include -L../../../submodules/Clipper2/emscripten_temp_install/lib -L../../../_build/libs/libmslicommon -L../../../_build/libs/libmslicommon/bin/muscle -DMUSCLE_ENABLE_ZLIB_ENCODING -lmslicommon -lmuscle -lClipper2 -lembind -sEXPORTED_RUNTIME_METHODS=ccall,cwrap,HEAPU8 -sWASM_BIGINT -sALLOW_MEMORY_GROWTH=1 -sMODULARIZE=1 -sEXPORT_ES6=1 -sEXPORT_NAME="CPPBindings" -sENVIRONMENT=web -lembind -lwebsocket.js -std=c++17 js_bindings.cpp -o cpp_bindings.js
In file included from js_bindings.cpp:1:
In file included from /Users/jaf/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/bind.h:24:
In file included from /Users/jaf/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/val.h:13:
/Users/jaf/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/wire.h:124:19: error: static assertion failed due to requirement '!std::is_pointer<JJObserved *>::value': Implicitly binding raw
      pointers is illegal. Specify allow_raw_pointer<arg<?>>
  124 |     static_assert(!std::is_pointer<T*>::value, "Implicitly binding raw pointers is illegal.  Specify allow_raw_pointer<arg<?>>");
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/jaf/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/wire.h:155:22: note: in instantiation of template class 'emscripten::internal::TypeID<JJObserved *>' requested here
  155 | struct TypeID<T&&> : TypeID<T> {
      |                      ^
/Users/jaf/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/wire.h:240:45: note: in instantiation of template class 'emscripten::internal::TypeID<JJObserved *&&>' requested here
  240 |         static constexpr TYPEID types[] = { TypeID<Args>::get()... };
      |                                             ^
/Users/jaf/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/wire.h:263:16: note: in instantiation of member function
      'emscripten::internal::ArgArrayGetter<emscripten::internal::TypeList<emscripten::internal::AllowedRawPointer<JJObserverWrapper>, emscripten::val &&, JJObserved *&&>>::get' requested here
  263 |             >::get();
      |                ^
/Users/jaf/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/bind.h:1735:18: note: in instantiation of member function
      'emscripten::internal::WithPolicies<emscripten::allow_raw_pointer<emscripten::ret_val>, emscripten::nonnull<emscripten::ret_val>>::ArgTypeList<JJObserverWrapper *, emscripten::val &&, JJObserved
      *&&>::getTypes' requested here
 1735 |             args.getTypes(),
      |                  ^
/Users/jaf/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/bind.h:1542:13: note: in instantiation of function template specialization
      'emscripten::class_<JJObserver>::class_function<JJObserverWrapper *, emscripten::val &&, JJObserved *&&, emscripten::allow_raw_pointer<emscripten::ret_val>, emscripten::nonnull<emscripten::ret_val>>'
      requested here
 1542 |             class_function(
      |             ^
js_bindings.cpp:34:10: note: in instantiation of function template specialization 'emscripten::class_<JJObserver>::allow_subclass<JJObserverWrapper, JJObserved *>' requested here
   34 |         .allow_subclass<JJObserverWrapper>("JJObserverWrapper", constructor<JJObserved *>());
      |          ^
/Users/jaf/emsdk/upstream/emscripten/cache/sysroot/include/emscripten/wire.h:332:13: note: unnamed type used in template argument was declared here
  332 |     typedef struct {
      |             ^
1 error generated.
emcc: error: '/Users/jaf/emsdk/upstream/bin/clang -target wasm32-unknown-emscripten -fignore-exceptions -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --sysroot=/Users/jaf/emsdk/upstream/emscripten/cache/sysroot -DEMSCRIPTEN -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -std=c++17 -c js_bindings.cpp -o /var/folders/q9/9w3p5x650wgfpbcws3zhsc6h0000gn/T/emscripten_temp_vwjyw3f2/js_bindings.o' failed (returned 1)
```

--- a/system/include/emscripten/bind.h
+++ b/system/include/emscripten/bind.h
@@ -1424,6 +1424,32 @@ struct RegisterClassMethod<FunctionTag<Callable, ReturnType (ThisType, Args...)>
     }
 };
 
+// Helper structs for shifting argument indices when a policy is applied.
+template<typename Slot>
+struct ShiftSlot {
+    using type = Slot;
+};
+
+template<int Index>
+struct ShiftSlot<arg<Index>> {
+    using type = arg<Index + 1>;
+};
+
+template<typename Policy>
+struct ShiftPolicy {
+    using type = Policy;
+};
+
+template<typename Slot>
+struct ShiftPolicy<allow_raw_pointer<Slot>> {
+    using type = allow_raw_pointer<typename ShiftSlot<Slot>::type>;
+};
+
+template<typename Slot>
+struct ShiftPolicy<nonnull<Slot>> {
+    using type = nonnull<typename ShiftSlot<Slot>::type>;
+};
+
 } // end namespace internal
 
 template<typename... ConstructorArgs>
@@ -1532,10 +1558,11 @@ public:
         return *this;
     }
 
-    template<typename WrapperType, typename... ConstructorArgs>
+    template<typename WrapperType, typename... ConstructorArgs, typename... Policies>
     EMSCRIPTEN_ALWAYS_INLINE const class_& allow_subclass(
         const char* wrapperClassName,
-        ::emscripten::constructor<ConstructorArgs...> = ::emscripten::constructor<>()
+        ::emscripten::constructor<ConstructorArgs...> = ::emscripten::constructor<>(),
+        Policies... policies
     ) const {
         using namespace internal;
 
@@ -1549,18 +1576,20 @@ public:
             class_function(
                 "implement",
                 &wrapped_new<WrapperType*, WrapperType, val, ConstructorArgs...>,
-                allow_raw_pointer<ret_val>(), nonnull<ret_val>())
+                allow_raw_pointer<ret_val>(), nonnull<ret_val>(),
+                typename ShiftPolicy<Policies>::type()...)
             .class_function(
                 "extend",
                 &wrapped_extend<WrapperType>)
             ;
     }
 
-    template<typename WrapperType, typename PointerType, typename... ConstructorArgs>
+    template<typename WrapperType, typename PointerType, typename... ConstructorArgs, typename... Policies>
     EMSCRIPTEN_ALWAYS_INLINE const class_& allow_subclass(
         const char* wrapperClassName,
         const char* pointerName,
-        ::emscripten::constructor<ConstructorArgs...> = ::emscripten::constructor<>()
+        ::emscripten::constructor<ConstructorArgs...> = ::emscripten::constructor<>(),
+        Policies... policies
     ) const {
         using namespace internal;
 
@@ -1575,7 +1604,8 @@ public:
             class_function(
                 "implement",
                 &wrapped_new<PointerType, WrapperType, val, ConstructorArgs...>,
-                allow_raw_pointer<ret_val>())
+                allow_raw_pointer<ret_val>(),
+                typename ShiftPolicy<Policies>::type()...)
             .class_function(
                 "extend",
                 &wrapped_extend<WrapperType>)

--- a/test/embind/test_embind_subclass_pointer.cpp
+++ b/test/embind/test_embind_subclass_pointer.cpp
@@ -1,0 +1,29 @@
+#include <emscripten/bind.h>
+
+using namespace emscripten;
+
+struct Base {
+    virtual ~Base() {}
+};
+
+struct Sub : Base {
+    Sub(Base* b) {}
+};
+
+struct SubWrapper : wrapper<Sub> {
+    EMSCRIPTEN_WRAPPER(SubWrapper);
+};
+
+EMSCRIPTEN_BINDINGS(test) {
+    class_<Base>("Base")
+        .constructor<>()
+        ;
+
+    class_<Sub>("Sub")
+        .allow_subclass<SubWrapper>("SubWrapper", constructor<Base*>(), allow_raw_pointer<arg<0>>())
+        ;
+}
+
+int main() {
+    return 0;
+}

--- a/test/test_other.py
+++ b/test/test_other.py
@@ -3387,6 +3387,9 @@ More info: https://emscripten.org
   def test_embind_allow_raw_pointer(self):
     self.emcc(test_file('embind/test_embind_allow_raw_pointer.cpp'), ['-lembind'])
 
+  def test_embind_subclass_pointer(self):
+    self.emcc(test_file('embind/test_embind_subclass_pointer.cpp'), ['-lembind'])
+
   @is_slow_test
   @parameterized({
     '': [],

